programs/fundamentals/matrix.sc

```diff
@@ -0,0 +1,176 @@
+                row = _;
+
+                if(length(matrix:_)!=col_size, throw()); //checking that all columns are the same length
+
+                loop(col_size,
+                    if(type(matrix:row:_)!='number',
+                        throw()
+                    )
+                )
+            ),
+
+            throw()
+        ),
+        return(false)
+    );
+    true
+);
+
+is_vector(vector)->//checking like this so the first check fails without having to iterate over the whole list potentially
+    !(type(vector)!='list' || length(vector)==0 || first(vector, type(_)!='number'));
+
+check_matrix(matrix)->( //checks if given value is a matrix, and if not throws an error
+    if(!is_matrix(matrix),// todo decide whether or not to give more detailed error, i.e what exactly went wrong
+        exit(print('Operation is invalid, "' + matrix + '" is not a matrix'))
+    )
+);
+
+
+identity(size)->(//returns identity matrix of given size
+    identity = [];
+    loop(size,
+        width = _;
+        identity:_=[];
+        loop(size, //cos its always a square matrix
+            identity:width:_ = width ==_
+        )
+    );
+    identity
+);
+
+get_column(matrix, column)->( //get_row func would be identical to matrix:row, whereas here you need that element of each list, which is O(n)
+    check_matrix(matrix);
+    if(length(matrix:0)<=column,
+        exit(print(str('Index out of bounds, cannot access %s column of %s by %s matrix', column, length(matrix), length(matrix:0))))
+    );
+    col = [];
+    loop(length(matrix),
+        col:_ = matrix:_:column
+    );
+    col
+);
+
+matrix_add(a, b) ->( //returns a new matrix whose values are the sum of the two input matrices
+    check_matrix(a);
+    check_matrix(b);
+
+    if(length(a)!=length(b) || length(a:0)!=length(b:0),
+        exit(print('Cannot add uneven matrices'))
+    );
+
+    sum = [];
+
+    loop(length(a),
+        width = _;
+        sum:width=[];
+        loop(length(a:0),
+            sum:width:_ = a:width:_ + b:width:_
+        )
+    );
+    sum
+);
+
+matrix_subtract(a, b) ->( //returns a new matrix whose values are the difference of the two input matrices (i.e a - b)
+    check_matrix(a);
+    check_matrix(b);
+
+    if(length(a)!=length(b) || length(a:0)!=length(b:0),
+        exit(print('Cannot subtract uneven matrices'))
+    );
+
+    sum = [];
+
+    loop(length(a),
+        width = _;
+        sum:width=[];
+        loop(length(a:0),
+            sum:width:_ = a:width:_ - b:width:_
+        )
+    );
+    sum
+);
+
+//This is complicated, because you can do many things here, depending if input is scalar, vector or matrix:
+//  scalar * scalar = error (y doin that op here)
+//  scalar * vector = error (y doin that op here)
+//  scalar * matrix = error (cannot do it in this order). Still gotta handle.
+//  vector * scalar = error (y doin that op here)
+//  vector * vector = matrix with rows same as length of first vector, columns same as that of second.
```

Comment:
This is called outer product of vectors, it's what happens when you multiply two "matrices" of sizes N×1 and 1×N, respectivelye (while the inverse 1×N times N×1 results in the usual dot product). If your "vector" objects are just gonna be lists, I'm not sure it's a good idea to include this behaviour in this function.

Put in another way, you need to decide what shape your vectors represent: is a list a 1×N (row vector) or a N×1 (column vector) object? In the latter case, outer product is `v×w transposed`, and `v transposed × w` is dot product.

Reply:
I think it makes more sense for vectors to be treated as 1*N matrices in this case, cos of the way they look in code: `[a, b, c]`, which looks more like a row vector than a column vector imo.

Reply:
I agree, although it's customary for vectors to default to N×1 matrices. Either way, you need ot make a choice, make it explicit and stick with it, imo.